service/internal/status/status.go

```diff
@@ -0,0 +1,167 @@
+				component.StatusPermanentError: {},
+				component.StatusFatalError:     {},
+				component.StatusStopping:       {},
+				component.StatusStopped:        {},
+			},
+			component.StatusPermanentError: {},
+			component.StatusFatalError:     {},
+			component.StatusStopping: {
+				component.StatusRecoverableError: {},
+				component.StatusPermanentError:   {},
+				component.StatusFatalError:       {},
+				component.StatusStopped:          {},
+			},
+			component.StatusStopped: {},
+		},
+	}
+}
+
+// InitFunc can be used to toggle a ready flag to true
+type InitFunc func()
+
+// readFunc can be used to check the value of a ready flag
+type readyFunc func() bool
+
+// initAndReadyFuncs returns a pair of functions to set and check a boolean ready flag
+func initAndReadyFuncs() (InitFunc, readyFunc) {
+	mu := sync.RWMutex{}
+	isReady := false
+
+	init := func() {
+		mu.Lock()
+		defer mu.Unlock()
+		isReady = true
+	}
+
+	ready := func() bool {
+		mu.RLock()
+		defer mu.RUnlock()
+		return isReady
+	}
+
+	return init, ready
+}
+
+// NotifyStatusFunc is the receiver of status events after successful state transitions
+type NotifyStatusFunc func(*component.InstanceID, *component.StatusEvent)
+
+// ServiceStatusFunc is the expected type of ReportComponentStatus for servicetelemetry.Settings
+type ServiceStatusFunc func(id *component.InstanceID, status component.Status, opts ...component.StatusEventOption) error
+
+// errStatusNotReady is returned when trying to report status before service start
+var errStatusNotReady = errors.New("report component status is not ready until service start")
+
+// NewServiceStatusFunc returns a function to be used as ReportComponentStatus for
+// servicetelemetry.Settings, which differs from component.TelemetrySettings in that
+// the service version does not correspond to a specific component, and thus needs
+// the a component.InstanceID as a parameter.
+func NewServiceStatusFunc(notifyStatusChange NotifyStatusFunc) (InitFunc, ServiceStatusFunc) {
+	init, isReady := initAndReadyFuncs()
+	// mu synchronizes access to the fsmMap and the underlying fsm during a state transition
+	mu := sync.Mutex{}
+	fsmMap := make(map[*component.InstanceID]*fsm)
+	return init,
+		func(id *component.InstanceID, status component.Status, opts ...component.StatusEventOption) error {
+			if !isReady() {
+				return errStatusNotReady
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			fsm, ok := fsmMap[id]
+			if !ok {
+				fsm = newFSM(func(ev *component.StatusEvent) {
+					notifyStatusChange(id, ev)
+				})
+				fsmMap[id] = fsm
+			}
+			return fsm.transition(status, opts...)
+		}
+
+}
+
+// NewComponentStatusFunc returns a function to be used as ReportComponentStatus for
+// component.TelemetrySettings, which differs from servicetelemetry.Settings in that
+// the component version is tied to specific component instance.
+func NewComponentStatusFunc(id *component.InstanceID, srvStatus ServiceStatusFunc) component.StatusFunc {
```

Comment:
One complication I want to call out again is the [`sharedcomponent`](https://github.com/open-telemetry/opentelemetry-collector/blob/2ebe1e501f9a206237de1009a127f1a419ab3696/internal/sharedcomponent/sharedcomponent.go) package. This can be used to combine instances together.

Take for instance the `otlp` receiver which uses this mechanism. The collector will believe that it is creating multiple instances, but the receiver implementation will use `sharedcomoponent` to check if any instance has already been created. If so, it will return that to the collector. The problem here is that whichever instance is created first will be the only one that ever reports status events.

I think it should be possible to fix this within the package. Basically, a shared component could hang on to the `StatusFunc` from each of the instances that it represents. Then whenever it reports an event, it should report it through all of the functions.

Reply:
I made an attempt at this. I modified `SharedComponent.GetOrAdd` to take a reference to a `component.TelemetrySettings` and modified the `ReportComponentStatus` of the `SharedComponent` to call the `ReportComponentStatus` methods of its "instances". I built the collector and checked that it works in the real world with the OTLP receiver.